File: Tags/CheckBoxController.vb
Option Explicit On
Imports System.Windows.Forms

Public Class CheckBoxController


    'By declaring Public WithEvents we can handle
    'events "collectively". In this case it is
    'the click event on a date label, and by
    'doing it this way we avoid writing click
    'events for each and every data label.
    Public WithEvents ctrlCB As CheckBox
    Public TrigByKeyChg As Boolean
    Private TrigByValChg As Boolean
    Private _parent As TagController
    Private strTagPrefix As String
    Private strTemp As String


    Friend Function Init(parent As TagController, strPrefix As String)
        _parent = parent
        strTagPrefix = strPrefix
        Return True
    End Function

    Private Sub ctrlCB_Click() Handles ctrlCB.Click
        If Not TrigByKeyChg Then
            strTemp = strTagPrefix & ctrlCB.Text
            _parent.ToggleChoice(strTemp)
            _parent.FocusCheckbox(ctrlCB)
        ElseIf TrigByValChg Then
            TrigByKeyChg = False
            TrigByValChg = False
        Else
            TrigByValChg = True
            ctrlCB.Checked = Not ctrlCB.Checked
        End If
        'Me.ctrlCB.Value = Not Me.ctrlCB.Value
    End Sub

    Private Sub ctrlCB_KeyDown(sender As Object, e As KeyEventArgs) Handles ctrlCB.KeyDown
        Select Case e.KeyCode
            Case Keys.Down
                _parent.Select_Ctrl_By_Offset(1)

            Case Keys.Up
                _parent.Select_Ctrl_By_Offset(-1)

            Case Keys.End
                _parent.Select_Last_Control()

            Case Keys.Home
                _parent.Select_First_Control()

            Case Keys.PageDown
                _parent.Select_PageDown()

            Case Keys.PageUp
                _parent.Select_PageUp()

            Case Keys.Enter
                _parent.OK_Action()
        End Select
    End Sub

    Private Sub ctrlCB_GotFocus(sender As Object, e As EventArgs) Handles ctrlCB.GotFocus
        Dim ctrl = TryCast(sender, Control)
        Dim tmp_color As Drawing.Color = ctrl.BackColor
        ctrl.BackColor = ctrl.ForeColor
        ctrl.ForeColor = tmp_color
    End Sub

    Private Sub ctrlCB_LostFocus(sender As Object, e As EventArgs) Handles ctrlCB.LostFocus
        Dim ctrl = TryCast(sender, Control)
        Dim tmp_color As Drawing.Color = ctrl.BackColor
        ctrl.BackColor = ctrl.ForeColor
        ctrl.ForeColor = tmp_color
    End Sub

    Private Sub ctrlCB_PreviewKeyDown(sender As Object, e As PreviewKeyDownEventArgs) Handles ctrlCB.PreviewKeyDown
        Select Case e.KeyCode
            Case Keys.Down
                e.IsInputKey = True
            Case Keys.Up
                e.IsInputKey = True
        End Select
    End Sub
End Class

File: Tags/CheckBoxController.cs
using System;
using System.Runtime.CompilerServices;
using System.Windows.Forms;

namespace Tags
{

    public class CheckBoxController
    {


        // By declaring Public WithEvents we can handle
        // events "collectively". In this case it is
        // the click event on a date label, and by
        // doing it this way we avoid writing click
        // events for each and every data label.
        private CheckBox _ctrlCB;

        public virtual CheckBox ctrlCB
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _ctrlCB;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_ctrlCB != null)
                {
                    _ctrlCB.Click -= (_, __) => ctrlCB_Click();
                    _ctrlCB.KeyDown -= ctrlCB_KeyDown;
                    _ctrlCB.GotFocus -= ctrlCB_GotFocus;
                    _ctrlCB.LostFocus -= ctrlCB_LostFocus;
                    _ctrlCB.PreviewKeyDown -= ctrlCB_PreviewKeyDown;
                }

                _ctrlCB = value;
                if (_ctrlCB != null)
                {
                    _ctrlCB.Click += (_, __) => ctrlCB_Click();
                    _ctrlCB.KeyDown += ctrlCB_KeyDown;
                    _ctrlCB.GotFocus += ctrlCB_GotFocus;
                    _ctrlCB.LostFocus += ctrlCB_LostFocus;
                    _ctrlCB.PreviewKeyDown += ctrlCB_PreviewKeyDown;
                }
            }
        }
        public bool TrigByKeyChg;
        private bool TrigByValChg;
        private TagController _parent;
        private string strTagPrefix;
        private string strTemp;

        //BUGFIX: CheckBoxController adds event to an object that is null
        public CheckBoxController()
        {
            ctrlCB.Click += (_, __) => ctrlCB_Click();
            ctrlCB.KeyDown += ctrlCB_KeyDown;
            ctrlCB.GotFocus += ctrlCB_GotFocus;
            ctrlCB.LostFocus += ctrlCB_LostFocus;
            ctrlCB.PreviewKeyDown += ctrlCB_PreviewKeyDown;
        }


        internal object Init(TagController parent, string strPrefix)
        {
            _parent = parent;
            strTagPrefix = strPrefix;
            return true;
        }

        private void ctrlCB_Click()
        {
            if (!TrigByKeyChg)
            {
                strTemp = strTagPrefix + ctrlCB.Text;
                _parent.ToggleChoice(strTemp);
                _parent.FocusCheckbox(ctrlCB);
            }
            else if (TrigByValChg)
            {
                TrigByKeyChg = false;
                TrigByValChg = false;
            }
            else
            {
                TrigByValChg = true;
                ctrlCB.Checked = !ctrlCB.Checked;
            }
            // Me.ctrlCB.Value = Not Me.ctrlCB.Value
        }

        private void ctrlCB_KeyDown(object sender, KeyEventArgs e)
        {
            switch (e.KeyCode)
            {
                case Keys.Down:
                    {
                        _parent.Select_Ctrl_By_Offset(1);
                        break;
                    }

                case Keys.Up:
                    {
                        _parent.Select_Ctrl_By_Offset(-1);
                        break;
                    }

                case Keys.End:
                    {
                        _parent.Select_Last_Control();
                        break;
                    }

                case Keys.Home:
                    {
                        _parent.Select_First_Control();
                        break;
                    }

                case Keys.PageDown:
                    {
                        _parent.Select_PageDown();
                        break;
                    }

                case Keys.PageUp:
                    {
                        _parent.Select_PageUp();
                        break;
                    }

                case Keys.Enter:
                    {
                        _parent.OK_Action();
                        break;
                    }
            }
        }

        private void ctrlCB_GotFocus(object sender, EventArgs e)
        {
            Control ctrl = sender as Control;
            var tmp_color = ctrl.BackColor;
            ctrl.BackColor = ctrl.ForeColor;
            ctrl.ForeColor = tmp_color;
        }

        private void ctrlCB_LostFocus(object sender, EventArgs e)
        {
            Control ctrl = sender as Control;
            var tmp_color = ctrl.BackColor;
            ctrl.BackColor = ctrl.ForeColor;
            ctrl.ForeColor = tmp_color;
        }

        private void ctrlCB_PreviewKeyDown(object sender, PreviewKeyDownEventArgs e)
        {
            switch (e.KeyCode)
            {
                case Keys.Down:
                    {
                        e.IsInputKey = true;
                        break;
                    }
                case Keys.Up:
                    {
                        e.IsInputKey = true;
                        break;
                    }
            }
        }
    }
}